fem/transfer.cpp

```diff
@@ -409,11 +608,213 @@ L2ProjectionGridTransfer::L2ProjectionL2Space::L2ProjectionL2Space(
+   const bool build_P = fes_lor.GetTrueVSize() >= fes_ho.GetTrueVSize();
+
+   // If the local mesh is empty, skip all computations
+   if (nel_ho == 0) { return; }
+
+   const CoarseFineTransformations &cf_tr = mesh_lor->GetRefinementTransforms();
+
+   int nref_max = 0;
+   Array<Geometry::Type> geoms;
+   mesh_ho->GetGeometries(mesh_ho->Dimension(), geoms);
+   for (int ig = 0; ig < geoms.Size(); ++ig)
+   {
+      Geometry::Type geom = geoms[ig];
+      nref_max = std::max(nref_max, cf_tr.point_matrices[geom].SizeK());
+   }
+
+   BuildHo2Lor(nel_ho, nel_lor, cf_tr);
+
+   offsets.SetSize(nel_ho+1);
+   offsets[0] = 0;
+   for (int iho = 0; iho < nel_ho; ++iho)
+   {
+      int nref = ho2lor.RowSize(iho);
+      const FiniteElement &fe_ho = *fes_ho.GetFE(iho);
+      const FiniteElement &fe_lor = *fes_lor.GetFE(ho2lor.GetRow(iho)[0]);
+      offsets[iho+1] = offsets[iho] + fe_ho.GetDof()*fe_lor.GetDof()*nref;
+   }
+
+   // R will contain the restriction (L^2 projection operator) defined on each
+   // coarse HO element (and corresponding patch of LOR elements)
+   R.SetSize(offsets[nel_ho]);
+
+   if (build_P)
+   {
+      // P will contain the corresponding prolongation operator
+      P.SetSize(offsets[nel_ho]);
+   }
+
+   // Assemble mixed mass matrix
+   Vector M_mixed_all;
+   MixedMassEA(fes_ho, fes_lor, M_mixed_all, d_mt);
+
+
+   // R = inv(M_L) * M_mixed
+   // Need to compute M_L
+   // Note: Using user-inputted M_LH IntegrationRule ir
+   // (higher order than needed) in order to re-use coeff
+   MassIntegrator mi;
+
+   Vector M_ea_lor;
+   int ndof_lor;
+   int ndof_ho;
+   int nref;
+   {
+      int iho = 0;
+      Array<int> lor_els;
+      ho2lor.GetRow(iho, lor_els);
+      nref = ho2lor.RowSize(iho);
+
+      const FiniteElement &fe_ho = *fes_ho.GetFE(iho);
+      const FiniteElement &fe_lor = *fes_lor.GetFE(lor_els[0]);
+      ndof_ho = fe_ho.GetDof();
+      ndof_lor = fe_lor.GetDof();
+
+      M_ea_lor.SetSize(ndof_lor*ndof_lor*nel_lor, d_mt);
+   }
+
+   const bool add = false;
+   mi.AssembleEA(fes_lor, M_ea_lor, add);
+
+   DenseTensor Minv_ear_lor;
+   Minv_ear_lor.SetSize(ndof_lor, ndof_lor, nel_lor, d_mt);
+   Minv_ear_lor.GetMemory().CopyFrom(M_ea_lor.GetMemory(), M_ea_lor.Size());
+
+   BatchedLinAlg::Invert(Minv_ear_lor);
+   {
+      // Recall mfem is column major
+      // ndof_lor x ndof_ho
+      auto v_M_mixed_all = Reshape(M_mixed_all.Read(), ndof_lor, ndof_ho, nref,
+                                   nel_ho);
+
+      // matrix is symmetric
+      auto v_Minv_ear_lor = Reshape(Minv_ear_lor.Read(), ndof_lor, ndof_lor,
+                                    nel_lor);
+
+      // ndof_lor x ndof_ho
+      auto v_R = Reshape(R.Write(), ndof_lor, nref, ndof_ho, nel_ho);
+
+      MFEM_VERIFY(nel_lor==nel_ho*nref, "nel_lor != nel_ho*nref");
+
+      // (ndofs_lor x ndofs_lor) x (ndofs_lor x ndof_ho)
+      mfem::forall_3D(nel_ho, ndof_lor, nref, ndof_ho, [=] MFEM_HOST_DEVICE (int iho)
```

Comment:
@artv3, this will crash with CUDA at higher orders because you will be requesting too many threads per block.

Since these kernels don't use shared memory, you can just use a simple 1D `mfem::forall`, and then extract the indices using quotient and remainder. See commit bd6c2ab61a1eb59adbffcf3adc6a412ed1f7c2bd for an example of how to do this change.

A better long-term strategy would be to permute the order of the indices here so this can be written as a batched matvec, and then you don't need to worry about this :)